src/App/IndexedName.cpp

```diff
@@ -0,0 +1,171 @@
+ *   Copyright (c) 2022 Zheng, Lei (realthunder) <realthunder.dev@gmail.com>*
+ *   Copyright (c) 2023 FreeCAD Project Association                         *
+ *                                                                          *
+ *   This file is part of FreeCAD.                                          *
+ *                                                                          *
+ *   FreeCAD is free software: you can redistribute it and/or modify it     *
+ *   under the terms of the GNU Lesser General Public License as            *
+ *   published by the Free Software Foundation, either version 2.1 of the   *
+ *   License, or (at your option) any later version.                        *
+ *                                                                          *
+ *   FreeCAD is distributed in the hope that it will be useful, but         *
+ *   WITHOUT ANY WARRANTY; without even the implied warranty of             *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU       *
+ *   Lesser General Public License for more details.                        *
+ *                                                                          *
+ *   You should have received a copy of the GNU Lesser General Public       *
+ *   License along with FreeCAD. If not, see                                *
+ *   <https://www.gnu.org/licenses/>.                                       *
+ *                                                                          *
+ ***************************************************************************/
+
+#include "PreCompiled.h"
+
+#ifndef _PreComp_
+# include <cstdlib>
+# include <unordered_set>
+#endif
+
+#include <QHash>
+
+#include "IndexedName.h"
+
+using namespace Data;
+
+struct ByteArray
+{
+    explicit ByteArray(QByteArray other)
+        :bytes(std::move(other))
+    {}
+
+    ByteArray(const ByteArray& other) = default;
+
+    ByteArray(ByteArray&& other) noexcept
+        :bytes(std::move(other.bytes))
+    {}
+
+    ~ByteArray() = default;
+
+    void ensureUnshared() const
+    {
+        QByteArray copy;
+        copy.append(bytes.constData(), bytes.size());
+        bytes = copy;
+    }
+
+    bool operator==(const ByteArray& other) const {
+        return bytes == other.bytes;
+    }
+
+    ByteArray &operator=(const ByteArray & other) {
+        bytes.clear();
+        bytes.append(other.bytes.constData(), other.bytes.size());
+        return *this;
+    }
+
+    ByteArray &operator= (ByteArray&& other) noexcept
+    {
+        bytes = std::move(other.bytes);
+        return *this;
+    }
+
+    mutable QByteArray bytes;
+};
+
+struct ByteArrayHasher
+{
+    std::size_t operator()(const ByteArray& bytes) const
+    {
+        return qHash(bytes.bytes);
+    }
+
+    std::size_t operator()(const QByteArray& bytes) const
+    {
+        return qHash(bytes);
+    }
+};
+
+inline bool isInvalidChar(char test)
+{
```

Comment:
Check this for potential problems with std::isalpha: https://en.cppreference.com/w/cpp/string/byte/isalpha
If you want to make sure to have a char in the range a-z or A-Z it might be better to do that directly.

Reply:
Is it the locale that might be the issue, or the missing cast to unsigned char? I don't mind switching back to the direct comparison, I was thinking that using isalpha made it clearer (and shorter), but it's a very minor thing.

Reply:
(I switched it to a direct comparison in d4d9ce7c18176d113713c330882ca4a10b68d55a).

Reply:
If locale is an issue I imagine `isalpha` or a `Qt::String` equivalent would be better since it might check beyond the English alphabet. Or this same property could make it worse if only a-zA-Z are to be allowed.

Reply:
The (unsigned)char is probably not a problem because the come from a string. But according to the example code isalpha() can theoretically be a problem when it allows characters that shouldn't be allowed like German umlauts.

But I have to admit that on my system the test code neither for C nor de_DE locale the `ß` was accepted.